2주차 ToDo REST API 만들기

euijeonghwang opened the pull request:
2주차 첫 PR 잘부탁드립니다~😀

euijeonghwang committed:
spring-boot-devtools 추가해라

서버 매번 껐다키는 대신 build만 다시 해서 테스트할 수 있도록 함

euijeonghwang committed:
Task model만들고 간단한 GET/POST 요청을 받아라

Task.java : Task 객체 만듦
TaskController.java :
  - GET /tasks 목록 리턴 완료.
  - POST /tasks 간단히 Task 객체 생성해서 Task List에 넣어줌
  - Id 1씩 증가되는 메서드 생성

euijeonghwang committed:
스프링 Annotation을 활용하여 GET/POST/PUT/PATCH/DELETE를 구현해라.

에러 없는 파라미터가 날라온다는 가정 하에 GET/POST/PUT/PATCH/DELETE 초기 구현 완료함

euijeonghwang committed:
@PutMapping과 @PatchMapping을 분리해라.

같이 두어도 컴파일 에러가 안나길래 되는 줄 알았으나, 런타임 에러가 발생하여 메서드 따로 분리함.

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,105 @@
+        return tasks.stream()
+                .filter(task -> task.getId().equals(id))
+                .findFirst()
+                .orElse(null);
```
여기서 null을 응답하면, 사용자에게 어떤 응답이 오나요? 스프링에서는 예외를 어떻게 처리하는지 배워보고 처리해 봅시다

https://spring.io/blog/2013/11/01/exception-handling-in-spring-mvc

euijeonghwang replied:
null을 응답할 경우 ResponseBody에 아무 값도 없지만, 200을 리턴합니다. orElseThrow를 사용하여 NPE를 던지게끔 수정했습니다~!

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,105 @@
+    @PutMapping("/{id}")
+    public Task putTask(@PathVariable(name = "id") Long id, @RequestBody Task task) {
+
+        Task filteredTask = tasks.stream()
+                            .filter(t -> t.getId().equals(id))
+                            .findFirst()
+                            .orElse(null);
+        filteredTask.setTitle(task.getTitle());
```
`filteredTask`가 null일 경우에 NPE가 발생할 수 있겠네요.

euijeonghwang replied:
null을 넣어주면 자동으로 NPE 발생할 줄 알았는데 `Optional`에 대해 좀 더 지식이 있어야할 것 같습니다,,!

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,105 @@
+        return filteredTask;
+    }
+
+    @PatchMapping("/{id}")
+    public Task patchTask(@PathVariable(name = "id") Long id, @RequestBody Task task) {
+
+        Task filteredTask = tasks.stream()
+                .filter(t -> t.getId().equals(id))
+                .findFirst()
+                .orElse(null);
+        filteredTask.setTitle(task.getTitle());
+
+        return filteredTask;
+    }
```
`PUT` 메서드 처리와 코드가 완전히 동일하네요. 하나의 메서드가 여러개의 메서드를 처리할 수 있을까요?

euijeonghwang replied:
`@RequestMapping`을 사용하여 method로 지정함으로써 공통화해주었습니다!

euijeonghwang committed:
@ExceptionHandler를 사용하여 예외를 처리하고, ErrorHandler 클래스를 만들어라.

@ExceptionHandler를 사용하여 각 에러에 해당하는 예외처리함.
ErrorHandler를 통해 Error code와 message만 사용자에게 보이도록 함.

euijeonghwang committed:
PUT과 PATCH api 메서드를 공통화해라.

RequestMapping의 method를 사용해 반복되던 메서드를 하나의 메서드로 취합함.

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,135 @@
+        return tasks.stream()
+                .filter(task -> task.getId().equals(id))
+                .findFirst()
+                .orElseThrow(() -> new NullPointerException("Task가 존재하지 않습니다. ID를 다시 한 번 확인해주세요. id="+id));
```
찾을 수 없어서 결과가 null이 되니 `NullPointerException`예외를 발생시키는 것 같아요. NullPointerException는 참조하는 대상이 null일 경우 발생하는 예외로, 여기서 예외인 경우는 주어진 id로 Task를 찾을 수 없을 때 인 것 같아요. 따라서 이 예외를 새롭게 정의한 후 처리해주는게 좋을 것 같습니다

euijeonghwang replied:
그렇군요, `TaskNotFoundException` 라는 사용자 정의 예외를 만들어서 사용하겠습니다!

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,135 @@
+                            .findFirst()
+                            .orElseThrow(() -> new NullPointerException("ID를 다시 한 번 확인해주세요. ID="+id));
+        //NPE 발생 가능
+        if (task.getTitle().isEmpty()) throw new NullPointerException("Title을 입력해주세요.");
```
사소하다고 생각될 수 있지만 한 줄 이라도 중괄호를 사용해 주는게 좋습니다. 이것 때문에 발생한 어마어마한 버그 사례가 있을 정도로, 실수하기 쉽게 만듭니다

hannut91 commented:
커밋 내용안에 `@`을 사용하면 다음과 같이 엉뚱한 사람을 태그할 수 있습니다.

![image](https://user-images.githubusercontent.com/14071105/202065823-0ac9071e-3cc7-4523-a5fb-ae6f98724db0.png)

따라서 (`) `@`을 사용할 때는 이 문자열로 감싼 형태로 작성하는 것이 좋습니다. '
```
`@ExceptionHandler`
```

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,135 @@
+    private List<Task> tasks = new ArrayList<>();
+    private Long id = 0L;
+
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    @ExceptionHandler(NullPointerException.class)
+    public ErrorHandler NPEHandler(NullPointerException ne) {
+        return new ErrorHandler("Not Found", ne.getMessage());
+    }
+
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    @ExceptionHandler(BadRequest.class)
+    public ErrorHandler BadRequestHandler(BadRequest badRequest) {
+        return new ErrorHandler("Bad Request", "다시 한 번 확인해주세요.");
+    }
+
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    @ExceptionHandler(HttpMessageNotReadableException.class)
+    public ErrorHandler notReadableHandler(HttpMessageNotReadableException he) {
+        return new ErrorHandler("Not Readable", "title이 존재하지 않습니다. 다시 한 번 확인해주세요.");
+    }
```
예외에 대해서 응답을 처리해 주셨군요. 다른 파일로 혹시 분리할 수 있을까요?

euijeonghwang committed:
예외에 대한 응답 처리들을 Task Controller와 분리해라.

TaskController 안에서 사용하던 예외 응답 처리 메서드들을 ExceptionAdvice라는 클래스로 따로 분리함.
ExceptionAdvice는 `@RestControllerAdivce` 애노테이션을 사용하여 전역에서 사용할 수 있게끔 해줌.

euijeonghwang committed:
Task를 찾을 수 없을 때 발생하는 예외 클래스를 정의하라

파라미터로 넘어오는 Id를 통해 Task를 찾을 수 없을 때 발생하는 예외를 사용자 정의 예외 클래스 생성함

euijeonghwang committed:
Id를 입력하는 path segment에 숫자가 아닌 다른 형식이 들어올 경우의 예외를 처리해라

ExceptionAdvice.java : MethodArgumentTypeMismatchException 예외 추가
TaskController.java  : 메서드 참조를 통한 람다식 간단히 표현

euijeonghwang committed:
Get에서 id가 없을 경우 TaskNotFoundException을 발생시켜라.

NPE 에서 TaskNotFoundException으로 변경

euijeonghwang committed:
List<Task> tasks를 final로 선언해라

final로 선언한다면, 다른 곳에서 new ArrayList<>()를 다시 할 수 없다. final 키워드를 붙여서 다른 곳에서 초기화를 할 수 없도록 한다.

euijeonghwang commented:
질문 있습니다.👋🏻 `ExceptionAdvice` 클래스를 지금처럼 각 에러마다 세세하게 나눠서 관리해야하는지, 아니면 `@ResponseStatus`에 따라 `@ExceptionHandler`에 에러 클래스들을 배열로 지정하여 코드를 조금이라도 줄여야 하는게 좋은가요? 그 기준을 모르겠습니다. 제가 생각하기에는 `ExceptionAdvice` 클래스가 좀 지저분하다는 느낌이 듭니다.

hannut91 commented:
에러가 점점 증가하면, `ExceptionAdevice`안에 코드들이 점점 늘어날까봐 걱정이신 것 같아요.

제 경험상으로는 예외에 있는 내용을 그대로 응답으로 쓰는 경우가 많이 있더라구요. 이런 경우는 공통으로 묶을 수 있는 것은 묶어서 하나로 처리할 수 있도록 했습니다. 그런데 항상 이런 경우 일수는 없죠. 예외가 발생한 것과 사용자에게 응답을 해야하는 것은 내용이 다를 수도 있기 때문에, 이런 경우에는 예외에 대해서 어떻게 응답할지 따로 처리를 해주었습니다

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,113 @@
+     * id 1씩 증가
+     * @return 1 증가된 id
+     */
+    private Long generateId() {
+        id += 1;
+        return id;
+    }
```
여기에 `synchronized` 키워드를 적용해야할까요? 적용해야 하는지 어떻게 실험할 수 있을까요

euijeonghwang replied:
이에 대한 테스트 코드는 작성하지 않았지만(못했지만), 여러명이 동시에 요청할 경우 id가 겹치는 상황이 있을 수 있다고 생각합니다. 따라서 `id += 1;`해주는 로직에 `synchronized`를 사용하였습니다. 스레드에 대해서 배우기만 해보고 적용한 적이 없어서 이에 대해 더 공부해야할 것 같습니다~

euijeonghwang replied:
스레드를 다시 공부하면서 궁금한 점이 생겼습니다. 
멀티 스레드를 구현하려면 `Runnable` 인터페이스나 `Thread` 클래스를 상속받아야 새로운 스레드를 생성을 하는데, 그렇다면 현재 애플리케이션에서는 `synchronized` 키워드가 필요 없는게 아닌가 싶습니다. 만약 `synchronized`를 사용해야 할 필요가 있다면 두 가지 상황 중 하나일 것 같습니다. 

- 동시에 유저들이 `POST` 할 수 있으므로 동기화를 적용해야 한다.
- 개발자가 아무런 스레드를 생성하지 않아도 JVM을 관리하기 위한 여러 스레드가 존재하므로 동기화를 적용해야 한다.

제가 제대로 이해한게 맞는지 질문드립니다~!

hannut91 replied:
필요한지 안한지 일단 잘 모르겠지만, 문제가 생길거야 라는 가설을 세워보겠습니다.

> 동시에 여러번 호출하면, id 생성에 문제가 생길것이다

라는 가설이겠죠. 그러면, 동시에 여러번 호출했을 때, 올바르게 생성되지 않은 것을 증명하면 가설이 맞는거겠죠. 그래서 다음과 같이 테스트를 살짝 고쳐서 동시에 100번을 한 번에 호출하게 한 번 해보겠습니다.

```js
 describe('POST /tasks', () => {
    it.only('responses 201 Created', async () => {
      const promise = [];
      for (let i = 1; i < 100; i++) {
        promise.push(
          frisby.post('/tasks', { title })
            .expect('status', 201),
        );
      }

      await Promise.all(promise);
    });
  });
```

그랬더니 생성된 결과가 다음과 같았습니다.

```json
[
    {
        "id": 1,
        "title": "책 읽기"
    },
    {
        "id": 1,
        "title": "책 읽기"
    },
    {
        "id": 2,
        "title": "책 읽기"
    },
    {
        "id": 1,
        "title": "책 읽기"
    },
    {
        "id": 3,
        "title": "책 읽기"
    },
    {
        "id": 4,
        "title": "책 읽기"
    },
    {
        "id": 5,
        "title": "책 읽기"
    },
    {
        "id": 6,
        "title": "책 읽기"
    },
    {
        "id": 7,
        "title": "책 읽기"
    },
    {
        "id": 8,
        "title": "책 읽기"
    },
    // ... 생략
]
```

가설이 맞았네요. 정상적으로 호출이 되지 않는 군요. 그러면 왜 그럴까? 를 이제 질문을 던져봐야 겠네요

euijeonghwang replied:
`generateId()`의 `id += 1`을 하고 있는 도중이거나 하기 전일 때 또 다른 `POST` 요청이 와서 `id`가 증가되기 전에 동시에 요청하다보니 id가 중복되는 현상이 발생합니다~!

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,113 @@
+@CrossOrigin
+public class TaskController {
+
+    private final List<Task> tasks = new ArrayList<>();
```
new ArrayList를 갓 만들었을 때 ArrayList의 capacity는 얼마인까요? 만약 잘 모르겠다면 이번 기회에 ArrayList 클래스를 열어서 조사해 봅시다.

euijeonghwang replied:
Default size는 10입니다. 하지만 ArrayList는 가변적인 자료구조이기 때문에, 10을 넘는다 하더라도 내부에서 자동으로 capacity를 늘려주게 됩니다. 하지만 ArrayList는 capacity를 int형으로 받기 때문에 max capacity는 32bit max integer인 2,147,483,647입니다. 이를 넘어갈 경우 LinkedList를 사용하면 될 것 같습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,113 @@
+        return tasks;
+    }
+
+    /**
+     * GET /tasks/{id}
+     * @param id
+     * @return task
+     */
```
Javadoc을 작성하기로 하셨군요. 사용하고 있는 아무 자바 라이브러리 중 하나를 보고 Javadoc을 어떻게 쓰고 있는지 살펴봅시다

euijeonghwang committed:
Id 생성 시 synchronized를 선언하여 thread safe하도록 해라.

Id를 생성하는 generateId() 메서드에 synchronized statements 방법으로 id를 1 증가하는 로직에만 synchronized를 선언함.
이때, lock 이라는 별도의 객체를 선언하여 사용함.

euijeonghwang committed:
Controller와 Service를 분리해라

Serivce 클래스를 만들어서 컨트롤러와 비즈니스 로직을 분리

euijeonghwang committed:
TaskService JavaDoc 작성해라

각 메서드마다 주석을 달아줌

hannut91 commented on app/src/main/java/com/codesoom/assignment/application/TaskService.java:
```diff
@@ -0,0 +1,95 @@
+    public Task updateTask(Long id, Task task) {
+        //NPE 발생 가능
+        if (task.getTitle().isEmpty()) {
+            throw new NullPointerException("Title을 입력해주세요.");
```
task가 비버있을 경우 NullPointerException이 발생해서 여기서 `NullPointerException`을 던지고 있는 것 같습니다. 왜 예외가 발생하는지 명확한 예외를 정의하는 것이 좋아 보입니다.

더 나아가서 `updateTask`에는 title이 비어있지 않은 요청이 오도록 Controller에서 검증을 해주면 좋겠네요

euijeonghwang replied:
위의 로직은 컨트롤러로 옮기고, 에러 메시지같은 경우 한 글자 이상 작성을 요구하도록 자세하게 적어주었습니다.

왜 예외가 발생하는지 명확한 예외를 정의한다는 말은 `NPE`가 아니라 다른 예외를 던지도록 해야한다는 말씀이신건가요? 찾아보니 `IllegalArgumentException` 을 던져야한다는 [답변](https://stackoverflow.com/questions/29366366/catch-empty-string-which-exception-to-use) 도 있어서 질문드립니다!

hannut91 replied:
IllegalArgumentException을 던져도 괜찮겠습니다. 근데 제가 생각하기에는 너무 구체적인 레벨의 예외라고 생각했어요. title이 null이 될 수 있다는 사실이 너무 낮은 레벨의 추상화 레벨을 가진 예외라고 생각했어요. 이 `updateTask`에서는 할 일을 업데이트하고, 업데이트 할 때 `할 일의 title이 없어요` 라고 얘기하는게 자연스럽다고 생각하는데, 갑자기 코드 레벨의 NullPointerException예외를 던진다고 생각했어요

euijeonghwang replied:
잘 이해가 안가는데, `NPE`를 던지기에는 과분하다는 뜻일까요?

hannut91 replied:
과분하다기 보다 이해하기 어렵다고 생각해요. NullPointerException이 발생했다는게 대체 무엇인가? 파악하기 어려울 것 같아요.

```java
public Task updateTask(Long id, Task task) throws NullPointerException {
    // ... 생략
}
```

위의 코드를 본다고 가정해 보면, `NullPointerException`를 던질 수도 있다는 사실이 잘 이해가 되시나요? 아니면

```java
public Task updateTask(Long id, Task task) throws TitleEmptyException {
    // ... 생략
}
```

이건 어떠세요?

euijeonghwang replied:
예제 보니 이해 됬어요~! 확실히 위의 코드는 id가 null이라는건지 task가 null이라는건지 구분이 안가지만 아래에 `TitleEmptyException`을 던져주면 title이 없다는 예외네요. ㅎㅎ 감사합니다

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,79 @@
+     */
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    public Task createList(@RequestBody Task task) {
```
`createTask`라는 이름은 어떨까요?

euijeonghwang replied:
`createTask`뿐만 아니라 `getLists`도 `getTasks`로 바꿔 반영했습니다

euijeonghwang committed:
메서드명을 Task와 관련되도록 수정하라

getLists 메서드를 getTasks로 수정함
createList 메서드를 createTask로 수정함

euijeonghwang committed:
Task update시 title 검증하는 로직을 Controller로 옮기고, 에러 메시지를 상세하게 적어라

RequestBody 자체가 넘어가지 않을 경우 ExceptionAdvice에서 notReadableExHandler()를 호출하고, RequestBody는 넘어가지만 title 값이 빈 값일 경우 컨트롤러에서 바로 NPE를 던져주도록 함.

hannut91 reviewed (approved):
이번 한 주도 고생 많으셨습니다
